cli/cli-client/src/main/java/org/infinispan/cli/patching/PatchTool.java

```diff
@@ -0,0 +1,474 @@
+                     }
+                  }
+                  break;
+               case ADD:
+               case HARD_REPLACE:
+               case UPGRADE:
+                  if (!dryRun) {
+                     Path file = target.resolve(operation.getNewPath());
+                     if (file.getParent() != null) {
+                        Files.createDirectories(file.getParent());
+                     }
+                     Files.copy(zipfs.getPath(operation.getNewPath().toString()), file, StandardCopyOption.REPLACE_EXISTING);
+                     Files.setPosixFilePermissions(file, PosixFilePermissions.fromString(operation.getNewPermissions()));
+                  }
+                  break;
+            }
+         }
+         patchInfo.setInstallationDate(new Date());
+         if (!dryRun) {
+            installedPatches.add(patchInfo);
+            writeInstalledPatches(target, installedPatches);
+         }
+         println(dryRun, MSG.patchInfo(patchInfo));
+      }
+   }
+
+   public void rollbackPatch(Path target, boolean dryRun) throws IOException {
+      List<PatchInfo> installedPatches = getInstalledPatches(target);
+      if (installedPatches.isEmpty()) {
+         throw MSG.patchNoPatchesInstalledToRollback();
+      }
+      PatchInfo patchInfo = installedPatches.remove(installedPatches.size() - 1);
+      Path backup = getBackupPath(target, patchInfo);
+      for (PatchOperation operation : patchInfo.getOperations()) {
+         switch (operation.getAction()) {
+            case ADD: {
+               Path file = target.resolve(operation.getNewPath());
+               // Remove any added files
+               println(dryRun, MSG.patchRollbackFile(file));
+               if (!dryRun) {
+                  Files.delete(file);
+               }
+            }
+            break;
+            case SOFT_REPLACE: {
+               // We only restore if the file hasn't been changed
+               Path file = target.resolve(operation.getPath());
+               String sha256 = sha256(file);
+               if (sha256 != null && sha256.equals(operation.getNewDigest())) {
+                  println(dryRun, MSG.patchRollbackFile(file));
+                  if (!dryRun) {
+                     Files.move(backup.resolve(operation.getPath()), target.resolve(operation.getPath()), StandardCopyOption.REPLACE_EXISTING);
+                  }
+               }
+               // We might have created a side-file, remove it
+               file = target.resolve(operation.getNewPath());
+               file = file.getParent().resolve(file.getFileName().toString() + "-" + patchInfo.getTargetVersion());
+               Files.deleteIfExists(file);
+               break;
+            }
+            case UPGRADE: {
+               Path file = target.resolve(operation.getNewPath());
+               println(dryRun, MSG.patchRollbackFile(file));
```

Comment:
When rollback, the upgrade files are printed twice:
```
Rolling back file '/tmp/redhat-datagrid-8.0.0-server/lib/protostream-4.4.0.CD20200427-redhat-00001.jar'
Rolling back file '/tmp/redhat-datagrid-8.0.0-server/lib/protostream-4.3.2.Final-redhat-00002.jar'
```
Not sure if intentional or not.